### PR TITLE
[TypeDeclaration] Handle string&string intersection on AddMethodCallBasedStrictParamTypeRector

### PR DESCRIPTION
**Pull request by samsonasik**
Given the following code:

```php
final class StringIntersection
{
    public function execute()
    {
        $value = sys_get_temp_dir() . DIRECTORY_SEPARATOR . basename(__FILE__);
        $this->run($value);
    }

    private function run($value)
    {
    }
}
```

It produce:

```diff
-    private function run($value)
+    private function run(string&string $value)
```

which will make Fatal error:

```bash
Fatal error: Type string cannot be part of an intersection type
```

ref https://3v4l.org/DLP7o#v8.1rc3

### CODE REVIEW

**Comment by samsonasik** [comment]
Fixed 🎉

**Comment by samsonasik** [comment]
All checks have passed 🎉 @TomasVotruba it is ready for review.

**Comment by samsonasik** [comment]
I updated to choose one if intersection type have same type

**Comment by samsonasik** [comment]
All checks have passed 🎉 @TomasVotruba it is ready for review.

File: rules/TypeDeclaration/NodeAnalyzer/ClassMethodParamTypeCompleter.php [hunk 1]
```diff
@@ -49,6 +52,18 @@ public function complete(ClassMethod $classMethod, array $classParameterTypes, i
...
+                $types = $phpParserTypeNode->types;
+
+                $firstType = (string) $this->nodeNameResolver->getName($types[0]);
+                $secondType = (string) $this->nodeNameResolver->getName($types[1]);
+
+                if ($firstType === $secondType) {
+                    // choose one
+                    $phpParserTypeNode = $types[0];
+                }
+            }
```
Comment (TomasVotruba) [inline rules/TypeDeclaration/NodeAnalyzer/ClassMethodParamTypeCompleter.php]: This looks weird. The intersectiontype should be probably narrowed to single one.

I'll check it locally

Comment (samsonasik) [reply rules/TypeDeclaration/NodeAnalyzer/ClassMethodParamTypeCompleter.php]: I moved check to IntersectionTypeMapper with check unique https://github.com/rectorphp/rector-src/pull/1375/commits/3cbb58ee08353f9a6753b59adff2a0fc665ab472

File: packages/NodeTypeResolver/NodeTypeResolver.php [hunk 2]
```diff
@@ -224,7 +224,8 @@ public function getNativeType(Expr $expr): Type
             return new MixedType();
         }
 
-        return $scope->getNativeType($expr);
+        $type = $scope->getNativeType($expr);
+        return $this->accessoryNonEmptyStringTypeCorrector->correct($type);
```
Comment (TomasVotruba) [inline packages/NodeTypeResolver/NodeTypeResolver.php:228]: Basically, this reduces `string|non-empty-string` to `string`. 
The `non-empty-string` is deteted by PHPStan, but not relevant for us, as all strict type declarations are only in `string` form.

**Pull request merged**